src/core/utils/geometryutils.cpp

```diff
@@ -94,24 +106,55 @@ GeometryUtils::GeometryOperationResult GeometryUtils::reshapeFromRubberband( Qgs
         case Qgis::AvoidIntersectionsMode::AllowIntersections:
           break;
       }
+
+      if ( QgsProject::instance()->topologicalEditing() )
+      {
+        // Since we're allowing reshaping in our logic when topological editing is on, remove the layer as it'll never overlap
+        avoidIntersectionsLayers.removeAll( layer );
+      }
```

Comment:
Or logic here was that when both avoid intersection and topological editing are on, we should reshape features on the layer of the targeted feature, this does just that.